### PR TITLE
Fix causal_lm cpp demo for llama architecture

### PR DESCRIPTION
**Pull request by sammysun0711**
This PR aim to fix causal_lm cpp demo for llama architecture
- Change directory name from `casual_lm` to `causal_lm`
- Fix causal lm cpp demo input & output mismatch for llama architecture
    - llama does not require position ids as input, only chatglm required
    - llama past kv cache as input starting from idx 2 instead of 3

This is a initial step to build baseline for consolidating following model support into causal_lm cpp demo:
- Qwen: https://github.com/openvinotoolkit/openvino.genai/pull/43
- ChatGLM: https://github.com/openvinotoolkit/openvino.genai/pull/48

### CODE REVIEW

**Comment by sammysun0711** [comment]
@Wovchena , @ilya-lavrenov, could you please review it?

File: text_generation/causal_lm/cpp/set_up_and_run.sh
Comment (Wovchena) [inline text_generation/causal_lm/cpp/set_up_and_run.sh:1]: Please, update https://github.com/openvinotoolkit/openvino.genai/blob/master/.github/workflows/casual_lm_cpp.yml as well

File: text_generation/causal_lm/cpp/set_up_and_run.sh
Comment (Wovchena) [inline text_generation/causal_lm/cpp/set_up_and_run.sh]: I'm fine if the test will pas. But as far as I remember that prompt results in infenit inference because EOS token is never produced.

File: text_generation/causal_lm/cpp/causal_lm.cpp
Comment (Wovchena) [inline text_generation/causal_lm/cpp/causal_lm.cpp:44]: > llama does not require position ids as input, only chatglm required

Converting it with optimum-cli as the README suggests, does generate position_ids

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: text_generation/causal_lm/cpp/set_up_and_run.sh
Comment (sammysun0711) [reply text_generation/causal_lm/cpp/set_up_and_run.sh:1]: Done

File: text_generation/causal_lm/cpp/set_up_and_run.sh
Comment (sammysun0711) [reply text_generation/causal_lm/cpp/set_up_and_run.sh]: Thanks, I will checkout locally first. 
May I know if this script is used only for CI check?

File: text_generation/causal_lm/cpp/causal_lm.cpp
Comment (ilya-lavrenov) [inline text_generation/causal_lm/cpp/causal_lm.cpp:47]: as far as I remember, chatglm has different index for batch dimension

File: text_generation/causal_lm/cpp/set_up_and_run.sh
Comment (Wovchena) [reply text_generation/causal_lm/cpp/set_up_and_run.sh]: Yes, only for CI. The idea was that people could reuse it. But later I extended tests with timeout: https://github.com/openvinotoolkit/openvino.genai/pull/38 and it became impossible to reuse it for group beam search

File: text_generation/causal_lm/cpp/causal_lm.cpp
Comment (sammysun0711) [reply text_generation/causal_lm/cpp/causal_lm.cpp:44]: Thanks for your info. We get llama OV model w/o position id with https://github.com/openvinotoolkit/openvino.genai/blob/master/llm_bench/python/convert.py
I suppose it is Optimum-Intel version issue. Let me checkout locally first.

File: text_generation/causal_lm/cpp/causal_lm.cpp
Comment (sammysun0711) [reply text_generation/causal_lm/cpp/causal_lm.cpp:47]: Yes, this PR is only for llama, will add chatglm support in another PR.

Comment (Wovchena) [reply text_generation/causal_lm/cpp/causal_lm.cpp:47]: Apparently exporting a model as stateful removes any difference between the dimensions. I was able to run chatglm2-6b and chatglm3-6bit using https://github.com/Wovchena/openvino.genai-public/pull/8 and get meaningful output. Here's the PR updating convert.py: https://github.com/openvinotoolkit/openvino.genai/pull/52. My next step is to rewrite this causal_lm as stateful as well, so we could decide if we need stateless approach at all.

Comment (ilya-lavrenov) [reply text_generation/causal_lm/cpp/causal_lm.cpp:47]: I agree.
But we should wait until stateful models support is merged to HF, right?
It would be great to avoid when we refer to different commits to convert different models

Comment (slyalin) [reply text_generation/causal_lm/cpp/causal_lm.cpp:47]: We are considering to have separate samples for stateless models if they demonstrate popular architectures and require architecture-dependent kv-cache processing.

Comment (slyalin) [reply text_generation/causal_lm/cpp/causal_lm.cpp:47]: But the main focus is on stateful models.

Comment (ilya-lavrenov) [reply text_generation/causal_lm/cpp/causal_lm.cpp]: so, maybe we can merge chatglm sample as is and treat it as "sample for stateless model for popular architecture" ?
While current one will show stateful models and hence it become more generic

Comment (Wovchena) [reply text_generation/causal_lm/cpp/causal_lm.cpp]: Greedy search with stateful model: https://github.com/sammysun0711/openvino.genai/pull/1 That eliminates a need to have special implementation for chatglm.

I also had that Idea to have stateless model with architecture, which doesn't fit into the general scenario. And there's still Qwen which may require that https://github.com/openvinotoolkit/openvino.genai/pull/43. I haven't tried it yet

File: text_generation/causal_lm/cpp/set_up_and_run.sh
Comment (sammysun0711) [reply text_generation/causal_lm/cpp/set_up_and_run.sh]: @Wovchena, I revert this change first to make CI green. 
BTW, do you think it is reasonable to add additional argument "max_nex_token" for loop condition: https://github.com/openvinotoolkit/openvino.genai/blob/0fae1f4c0fbe37dab68f35edcfbe4a1d0ac35742/text_generation/casual_lm/cpp/casual_lm.cpp#L72,  and this arguments can be pass by cmd line, which can better handle timeout cause by "EOS token never produced" issue.

File: text_generation/causal_lm/cpp/causal_lm.cpp
Comment (sammysun0711) [reply text_generation/causal_lm/cpp/causal_lm.cpp]: I think new optimum version add position_id as model input for onnx export: https://github.com/huggingface/optimum/pull/1381
@eaidova, could you add some comments for optimum-intel openvino export path? Or you recommend to use https://github.com/openvinotoolkit/openvino.genai/blob/master/llm_bench/python/convert.py for model conversion?

**Comment by sammysun0711** [comment]
@Wovchena, @ilya-lavrenov , @slyalin thanks for your feedback. I've learn a lot from it. 
Based on discussion above,  I think I will change this PR scope with only following changes: 
- Fix directory name typo 
- Upgrade OpenVINO 23.2 in https://github.com/openvinotoolkit/openvino.genai/blob/master/text_generation/casual_lm/cpp/set_up_and_run.sh for CI

As Ilya suggested, we can keep causal_lm sample for generic stateful model enabling. 
Meanwhile,  can we add ChatGLM3 and Qwen as stateless sample that demonstrate popular architectures and require architecture-dependent kv-cache processing.
- Qwen: https://github.com/openvinotoolkit/openvino.genai/pull/43 (Will update it with OV tokenizer)
- ChatGLM: https://github.com/openvinotoolkit/openvino.genai/pull/48

Please feel free to share any comments and feedback, thanks!

File: text_generation/causal_lm/cpp/causal_lm.cpp
Comment (sammysun0711) [reply text_generation/causal_lm/cpp/causal_lm.cpp]: Revert to use position ids for new optimum used in CI test

File: .github/workflows/casual_lm_cpp.yml
Comment (ilya-lavrenov) [inline .github/workflows/casual_lm_cpp.yml:1]: please, rename workflow file as well.

Comment (sammysun0711) [reply .github/workflows/casual_lm_cpp.yml:1]: Updated

**Review by ilya-lavrenov — APPROVED** [review]

File: text_generation/causal_lm/cpp/set_up_and_run.sh
Comment (Wovchena) [reply text_generation/causal_lm/cpp/set_up_and_run.sh]: > EOS token never produced" issue

is a CI issue and there's a way to solve it - use different prompt

**Pull request merged**
